File: libcudf-sys/src/ast.cpp
#include "ast.h"

#include <cudf/fixed_point/fixed_point.hpp>
#include <cudf/scalar/scalar.hpp>
#include <stdexcept>

namespace libcudf_bridge {

static_assert(static_cast<int32_t>(cudf::ast::table_reference::LEFT) == 0);
static_assert(static_cast<int32_t>(cudf::ast::table_reference::RIGHT) == 1);
static_assert(static_cast<int32_t>(cudf::ast::table_reference::OUTPUT) == 2);

#define LIBCUDF_ASSERT_AST_OPERATOR(name, value) \
    static_assert(static_cast<int32_t>(cudf::ast::ast_operator::name) == value)

LIBCUDF_ASSERT_AST_OPERATOR(ADD, 0);
LIBCUDF_ASSERT_AST_OPERATOR(SUB, 1);
LIBCUDF_ASSERT_AST_OPERATOR(MUL, 2);
LIBCUDF_ASSERT_AST_OPERATOR(DIV, 3);
LIBCUDF_ASSERT_AST_OPERATOR(TRUE_DIV, 4);
LIBCUDF_ASSERT_AST_OPERATOR(FLOOR_DIV, 5);
LIBCUDF_ASSERT_AST_OPERATOR(MOD, 6);
LIBCUDF_ASSERT_AST_OPERATOR(PYMOD, 7);
LIBCUDF_ASSERT_AST_OPERATOR(POW, 8);
LIBCUDF_ASSERT_AST_OPERATOR(EQUAL, 9);
LIBCUDF_ASSERT_AST_OPERATOR(NULL_EQUAL, 10);
LIBCUDF_ASSERT_AST_OPERATOR(NOT_EQUAL, 11);
LIBCUDF_ASSERT_AST_OPERATOR(LESS, 12);
LIBCUDF_ASSERT_AST_OPERATOR(GREATER, 13);
LIBCUDF_ASSERT_AST_OPERATOR(LESS_EQUAL, 14);
LIBCUDF_ASSERT_AST_OPERATOR(GREATER_EQUAL, 15);
LIBCUDF_ASSERT_AST_OPERATOR(BITWISE_AND, 16);
LIBCUDF_ASSERT_AST_OPERATOR(BITWISE_OR, 17);
LIBCUDF_ASSERT_AST_OPERATOR(BITWISE_XOR, 18);
LIBCUDF_ASSERT_AST_OPERATOR(LOGICAL_AND, 19);
LIBCUDF_ASSERT_AST_OPERATOR(NULL_LOGICAL_AND, 20);
LIBCUDF_ASSERT_AST_OPERATOR(LOGICAL_OR, 21);
LIBCUDF_ASSERT_AST_OPERATOR(NULL_LOGICAL_OR, 22);
LIBCUDF_ASSERT_AST_OPERATOR(IDENTITY, 23);
LIBCUDF_ASSERT_AST_OPERATOR(IS_NULL, 24);
LIBCUDF_ASSERT_AST_OPERATOR(SIN, 25);
LIBCUDF_ASSERT_AST_OPERATOR(COS, 26);
LIBCUDF_ASSERT_AST_OPERATOR(TAN, 27);
LIBCUDF_ASSERT_AST_OPERATOR(ARCSIN, 28);
LIBCUDF_ASSERT_AST_OPERATOR(ARCCOS, 29);
LIBCUDF_ASSERT_AST_OPERATOR(ARCTAN, 30);
LIBCUDF_ASSERT_AST_OPERATOR(SINH, 31);
LIBCUDF_ASSERT_AST_OPERATOR(COSH, 32);
LIBCUDF_ASSERT_AST_OPERATOR(TANH, 33);
LIBCUDF_ASSERT_AST_OPERATOR(ARCSINH, 34);
LIBCUDF_ASSERT_AST_OPERATOR(ARCCOSH, 35);
LIBCUDF_ASSERT_AST_OPERATOR(ARCTANH, 36);
LIBCUDF_ASSERT_AST_OPERATOR(EXP, 37);
LIBCUDF_ASSERT_AST_OPERATOR(LOG, 38);
LIBCUDF_ASSERT_AST_OPERATOR(SQRT, 39);
LIBCUDF_ASSERT_AST_OPERATOR(CBRT, 40);
LIBCUDF_ASSERT_AST_OPERATOR(CEIL, 41);
LIBCUDF_ASSERT_AST_OPERATOR(FLOOR, 42);
LIBCUDF_ASSERT_AST_OPERATOR(ABS, 43);
LIBCUDF_ASSERT_AST_OPERATOR(RINT, 44);
LIBCUDF_ASSERT_AST_OPERATOR(BIT_INVERT, 45);
LIBCUDF_ASSERT_AST_OPERATOR(NOT, 46);
LIBCUDF_ASSERT_AST_OPERATOR(CAST_TO_INT64, 47);
LIBCUDF_ASSERT_AST_OPERATOR(CAST_TO_UINT64, 48);
LIBCUDF_ASSERT_AST_OPERATOR(CAST_TO_FLOAT64, 49);

#undef LIBCUDF_ASSERT_AST_OPERATOR

namespace {
template <typename ScalarType>
std::size_t add_literal(
    AstExpressionTree& tree,
    const Scalar& scalar,
    const cudf::type_id expected_type)
{
    if (!scalar.inner) {
        throw std::runtime_error("Cannot add null scalar to AST expression tree");
    }
    if (scalar.inner->type().id() != expected_type) {
        throw std::invalid_argument("Scalar type does not match AST literal constructor");
    }
    // cuDF's literal constructors take non-const references but retain only const
    // references. The bridge keeps the wrapper const and borrows its stable allocation.
    tree.inner.emplace<cudf::ast::literal>(static_cast<ScalarType&>(*scalar.inner));
    return tree.inner.size() - 1;
}
} // namespace

AstExpressionTree::AstExpressionTree() = default;

AstExpressionTree::~AstExpressionTree() = default;

std::unique_ptr<AstExpressionTree> ast_expression_tree_create()
{
    return std::make_unique<AstExpressionTree>();
}

std::size_t ast_expression_tree_add_column_reference(
    AstExpressionTree& tree,
    int32_t column_index,
    int32_t table_reference)
{
    tree.inner.emplace<cudf::ast::column_reference>(
        column_index,
        static_cast<cudf::ast::table_reference>(table_reference));
    return tree.inner.size() - 1;
}

std::size_t ast_expression_tree_add_column_name_reference(
    AstExpressionTree& tree,
    rust::Str column_name)
{
    tree.inner.emplace<cudf::ast::column_name_reference>(
        std::string(column_name.data(), column_name.size()));
    return tree.inner.size() - 1;
}

#define LIBCUDF_DEFINE_AST_LITERAL(name, scalar_type, expected_id)                 \
    std::size_t ast_expression_tree_add_literal_##name(                            \
        AstExpressionTree& tree, const Scalar& scalar)                             \
    {                                                                               \
        return add_literal<scalar_type>(tree, scalar, cudf::type_id::expected_id);  \
    }

LIBCUDF_DEFINE_AST_LITERAL(int8, cudf::numeric_scalar<int8_t>, INT8)
LIBCUDF_DEFINE_AST_LITERAL(int16, cudf::numeric_scalar<int16_t>, INT16)
LIBCUDF_DEFINE_AST_LITERAL(int32, cudf::numeric_scalar<int32_t>, INT32)
LIBCUDF_DEFINE_AST_LITERAL(int64, cudf::numeric_scalar<int64_t>, INT64)
LIBCUDF_DEFINE_AST_LITERAL(uint8, cudf::numeric_scalar<uint8_t>, UINT8)
LIBCUDF_DEFINE_AST_LITERAL(uint16, cudf::numeric_scalar<uint16_t>, UINT16)
LIBCUDF_DEFINE_AST_LITERAL(uint32, cudf::numeric_scalar<uint32_t>, UINT32)
LIBCUDF_DEFINE_AST_LITERAL(uint64, cudf::numeric_scalar<uint64_t>, UINT64)
LIBCUDF_DEFINE_AST_LITERAL(float32, cudf::numeric_scalar<float>, FLOAT32)
LIBCUDF_DEFINE_AST_LITERAL(float64, cudf::numeric_scalar<double>, FLOAT64)
LIBCUDF_DEFINE_AST_LITERAL(bool8, cudf::numeric_scalar<bool>, BOOL8)
LIBCUDF_DEFINE_AST_LITERAL(string, cudf::string_scalar, STRING)
LIBCUDF_DEFINE_AST_LITERAL(timestamp_days, cudf::timestamp_scalar<cudf::timestamp_D>, TIMESTAMP_DAYS)
LIBCUDF_DEFINE_AST_LITERAL(timestamp_seconds, cudf::timestamp_scalar<cudf::timestamp_s>, TIMESTAMP_SECONDS)
LIBCUDF_DEFINE_AST_LITERAL(timestamp_milliseconds, cudf::timestamp_scalar<cudf::timestamp_ms>, TIMESTAMP_MILLISECONDS)
LIBCUDF_DEFINE_AST_LITERAL(timestamp_microseconds, cudf::timestamp_scalar<cudf::timestamp_us>, TIMESTAMP_MICROSECONDS)
LIBCUDF_DEFINE_AST_LITERAL(timestamp_nanoseconds, cudf::timestamp_scalar<cudf::timestamp_ns>, TIMESTAMP_NANOSECONDS)
LIBCUDF_DEFINE_AST_LITERAL(duration_days, cudf::duration_scalar<cudf::duration_D>, DURATION_DAYS)
LIBCUDF_DEFINE_AST_LITERAL(duration_seconds, cudf::duration_scalar<cudf::duration_s>, DURATION_SECONDS)
LIBCUDF_DEFINE_AST_LITERAL(duration_milliseconds, cudf::duration_scalar<cudf::duration_ms>, DURATION_MILLISECONDS)
LIBCUDF_DEFINE_AST_LITERAL(duration_microseconds, cudf::duration_scalar<cudf::duration_us>, DURATION_MICROSECONDS)
LIBCUDF_DEFINE_AST_LITERAL(duration_nanoseconds, cudf::duration_scalar<cudf::duration_ns>, DURATION_NANOSECONDS)
LIBCUDF_DEFINE_AST_LITERAL(decimal32, cudf::fixed_point_scalar<numeric::decimal32>, DECIMAL32)
LIBCUDF_DEFINE_AST_LITERAL(decimal64, cudf::fixed_point_scalar<numeric::decimal64>, DECIMAL64)
LIBCUDF_DEFINE_AST_LITERAL(decimal128, cudf::fixed_point_scalar<numeric::decimal128>, DECIMAL128)

#undef LIBCUDF_DEFINE_AST_LITERAL

std::size_t ast_expression_tree_add_operation(
    AstExpressionTree& tree,
    int32_t ast_operator,
    std::size_t left_index,
    std::size_t right_index)
{
    tree.inner.emplace<cudf::ast::operation>(
        static_cast<cudf::ast::ast_operator>(ast_operator),
        tree.inner.at(left_index),
        tree.inner.at(right_index));
    return tree.inner.size() - 1;
}

std::size_t ast_expression_tree_add_unary_operation(
    AstExpressionTree& tree,
    int32_t ast_operator,
    std::size_t input_index)
{
    tree.inner.emplace<cudf::ast::operation>(
        static_cast<cudf::ast::ast_operator>(ast_operator),
        tree.inner.at(input_index));
    return tree.inner.size() - 1;
}

} // namespace libcudf_bridge

File: src/ast.rs
use crate::{CuDFError, CuDFScalar};
use arrow::error::ArrowError;
use cxx::UniquePtr;
use libcudf_sys::{ffi, AstOperator, AstTableReference, TypeId};
use std::fmt;

/// A node index inside a [`CuDFAstExpression`] tree.
pub type CuDFAstNode = usize;

/// Table side used by a cuDF AST column reference.
///
/// Join predicates evaluate against two input tables. `Left` and `Right`
/// select which table a column reference reads from.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum CuDFAstTableReference {
    /// Column index in the left table.
    Left,
    /// Column index in the right table.
    Right,
    /// Column index in the output table.
    Output,
}

impl CuDFAstTableReference {
    fn into_sys(self) -> AstTableReference {
        match self {
            Self::Left => AstTableReference::Left,
            Self::Right => AstTableReference::Right,
            Self::Output => AstTableReference::Output,
        }
    }
}

/// Operators supported by [`CuDFAstExpression`].
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum CuDFAstOperator {
    /// Addition.
    Add,
    /// Subtraction.
    Sub,
    /// Multiplication.
    Mul,
    /// Division.
    Div,
    /// Equality comparison.
    Equal,
    /// Non-equality comparison.
    NotEqual,
    /// Less-than comparison.
    Less,
    /// Greater-than comparison.
    Greater,
    /// Less-than-or-equal comparison.
    LessEqual,
    /// Greater-than-or-equal comparison.
    GreaterEqual,
    /// Logical AND.
    LogicalAnd,
    /// Null-aware logical AND.
    NullLogicalAnd,
    /// Logical OR.
    LogicalOr,
    /// Null-aware logical OR.
    NullLogicalOr,
    /// Modulo.
    Mod,
    /// Null-aware equality comparison.
    NullEqual,
    /// Null check.
    IsNull,
    /// Logical NOT.
    Not,
    /// Cast to int64.
    CastToInt64,
    /// Cast to uint64.
    CastToUint64,
    /// Cast to float64.
    CastToFloat64,
}

impl CuDFAstOperator {
    fn into_sys(self) -> AstOperator {
        match self {
            Self::Add => AstOperator::Add,
            Self::Sub => AstOperator::Sub,
            Self::Mul => AstOperator::Mul,
            Self::Div => AstOperator::Div,
            Self::Equal => AstOperator::Equal,
            Self::NotEqual => AstOperator::NotEqual,
            Self::Less => AstOperator::Less,
            Self::Greater => AstOperator::Greater,
            Self::LessEqual => AstOperator::LessEqual,
            Self::GreaterEqual => AstOperator::GreaterEqual,
            Self::LogicalAnd => AstOperator::LogicalAnd,
            Self::NullLogicalAnd => AstOperator::NullLogicalAnd,
            Self::LogicalOr => AstOperator::LogicalOr,
            Self::NullLogicalOr => AstOperator::NullLogicalOr,
            Self::Mod => AstOperator::Mod,
            Self::NullEqual => AstOperator::NullEqual,
            Self::IsNull => AstOperator::IsNull,
            Self::Not => AstOperator::Not,
            Self::CastToInt64 => AstOperator::CastToInt64,
            Self::CastToUint64 => AstOperator::CastToUint64,
            Self::CastToFloat64 => AstOperator::CastToFloat64,
        }
    }
}

/// Owning cuDF AST expression tree.
///
/// Literal scalars are kept alive by this wrapper because cuDF AST literal
/// nodes reference scalar objects owned outside the tree.
///
/// The most recently added node is the expression root used by join filtering.
pub struct CuDFAstExpression {
    inner: UniquePtr<ffi::AstExpressionTree>,
    literals: Vec<CuDFScalar>,
}

impl fmt::Debug for CuDFAstExpression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CuDFAstExpression").finish_non_exhaustive()
    }
}

impl CuDFAstExpression {
    /// Create an empty cuDF AST expression tree.
    pub fn new() -> Self {
        Self {
            inner: ffi::ast_expression_tree_create(),
            literals: Vec::new(),
        }
    }

    /// Add a column reference node.
    ///
    /// `column_index` is relative to the table selected by `table`. For join
    /// filters this is usually a projected conditional table, not necessarily
    /// the full input table.
    ///
    /// # Errors
    ///
    /// Returns an error if `column_index` cannot be represented as a cuDF
    /// column index.
    pub fn column_reference(
        &mut self,
        column_index: usize,
        table: CuDFAstTableReference,
    ) -> Result<CuDFAstNode, CuDFError> {
        Ok(ffi::ast_expression_tree_add_column_reference(
            self.inner.pin_mut(),
            column_index.try_into().map_err(|_| {
                CuDFError::ArrowError(ArrowError::InvalidArgumentError(format!(
                    "AST column index {column_index} exceeds i32"
                )))
            })?,
            table.into_sys() as i32,
        )?)
    }

    /// Add a column-name reference node.
    ///
    /// cuDF Parquet filters use names so filter columns do not need to be
    /// present in the projected output.
    pub fn column_name_reference(
        &mut self,
        column_name: impl AsRef<str>,
    ) -> Result<CuDFAstNode, CuDFError> {
        Ok(ffi::ast_expression_tree_add_column_name_reference(
            self.inner.pin_mut(),
            column_name.as_ref(),
        )?)
    }

    /// Add a literal node.
    ///
    /// The scalar is moved into this expression tree wrapper and kept alive for
    /// as long as the AST exists.
    ///
    /// # Errors
    ///
    /// Returns an error if cuDF cannot add the scalar as an AST literal.
    pub fn literal(&mut self, scalar: CuDFScalar) -> Result<CuDFAstNode, CuDFError> {
        let dtype = scalar.inner().data_type()?.id();
        let tree = self.inner.pin_mut();
        let inner = scalar.inner();
        // SAFETY: the selected wrapper matches the scalar type id read from the
        // same cuDF scalar, and `self.literals` retains its allocation for the tree.
        // The C++ wrapper validates the type again before casting.
        let node = unsafe {
            match dtype {
                x if x == TypeId::Int8 as i32 => {
                    ffi::ast_expression_tree_add_literal_int8(tree, inner)
                }
                x if x == TypeId::Int16 as i32 => {
                    ffi::ast_expression_tree_add_literal_int16(tree, inner)
                }
                x if x == TypeId::Int32 as i32 => {
                    ffi::ast_expression_tree_add_literal_int32(tree, inner)
                }
                x if x == TypeId::Int64 as i32 => {
                    ffi::ast_expression_tree_add_literal_int64(tree, inner)
                }
                x if x == TypeId::Uint8 as i32 => {
                    ffi::ast_expression_tree_add_literal_uint8(tree, inner)
                }
                x if x == TypeId::Uint16 as i32 => {
                    ffi::ast_expression_tree_add_literal_uint16(tree, inner)
                }
                x if x == TypeId::Uint32 as i32 => {
                    ffi::ast_expression_tree_add_literal_uint32(tree, inner)
                }
                x if x == TypeId::Uint64 as i32 => {
                    ffi::ast_expression_tree_add_literal_uint64(tree, inner)
                }
                x if x == TypeId::Float32 as i32 => {
                    ffi::ast_expression_tree_add_literal_float32(tree, inner)
                }
                x if x == TypeId::Float64 as i32 => {
                    ffi::ast_expression_tree_add_literal_float64(tree, inner)
                }
                x if x == TypeId::Bool8 as i32 => {
                    ffi::ast_expression_tree_add_literal_bool8(tree, inner)
                }
                x if x == TypeId::String as i32 => {
                    ffi::ast_expression_tree_add_literal_string(tree, inner)
                }
                x if x == TypeId::TimestampDays as i32 => {
                    ffi::ast_expression_tree_add_literal_timestamp_days(tree, inner)
                }
                x if x == TypeId::TimestampSeconds as i32 => {
                    ffi::ast_expression_tree_add_literal_timestamp_seconds(tree, inner)
                }
                x if x == TypeId::TimestampMilliseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_timestamp_milliseconds(tree, inner)
                }
                x if x == TypeId::TimestampMicroseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_timestamp_microseconds(tree, inner)
                }
                x if x == TypeId::TimestampNanoseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_timestamp_nanoseconds(tree, inner)
                }
                x if x == TypeId::DurationDays as i32 => {
                    ffi::ast_expression_tree_add_literal_duration_days(tree, inner)
                }
                x if x == TypeId::DurationSeconds as i32 => {
                    ffi::ast_expression_tree_add_literal_duration_seconds(tree, inner)
                }
                x if x == TypeId::DurationMilliseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_duration_milliseconds(tree, inner)
                }
                x if x == TypeId::DurationMicroseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_duration_microseconds(tree, inner)
                }
                x if x == TypeId::DurationNanoseconds as i32 => {
                    ffi::ast_expression_tree_add_literal_duration_nanoseconds(tree, inner)
                }
                x if x == TypeId::Decimal32 as i32 => {
                    ffi::ast_expression_tree_add_literal_decimal32(tree, inner)
                }
                x if x == TypeId::Decimal64 as i32 => {
                    ffi::ast_expression_tree_add_literal_decimal64(tree, inner)
                }
                x if x == TypeId::Decimal128 as i32 => {
                    ffi::ast_expression_tree_add_literal_decimal128(tree, inner)
                }
                _ => {
                    return Err(ArrowError::NotYetImplemented(format!(
                        "cuDF AST literals do not support type id {dtype}"
                    ))
                    .into())
                }
            }
        }?;
        self.literals.push(scalar);
        Ok(node)
    }

    /// Add a unary operation node.
    ///
    /// # Errors
    ///
    /// Returns an error if `input` does not refer to an existing AST node or
    /// cuDF rejects the requested operation.
    pub fn unary_operation(
        &mut self,
        op: CuDFAstOperator,
        input: CuDFAstNode,
    ) -> Result<CuDFAstNode, CuDFError> {
        Ok(ffi::ast_expression_tree_add_unary_operation(
            self.inner.pin_mut(),
            op.into_sys() as i32,
            input,
        )?)
    }

    /// Add a binary operation node.
    ///
    /// # Errors
    ///
    /// Returns an error if either operand does not refer to an existing AST
    /// node or cuDF rejects the requested operation.
    pub fn binary_operation(
        &mut self,
        op: CuDFAstOperator,
        left: CuDFAstNode,
        right: CuDFAstNode,
    ) -> Result<CuDFAstNode, CuDFError> {
        Ok(ffi::ast_expression_tree_add_operation(
            self.inner.pin_mut(),
            op.into_sys() as i32,
            left,
            right,
        )?)
    }

    pub(crate) fn inner(&self) -> &UniquePtr<ffi::AstExpressionTree> {
        &self.inner
    }
}

impl Default for CuDFAstExpression {
    fn default() -> Self {
        Self::new()
    }
}
